tmt/steps/__init__.py

```diff
@@ -345,37 +364,85 @@ def usage(self) -> str:
         return re.sub('\n\n', '\n\n\b\n', usage)
 
 
+class PluginData(TypedDict, total=False):
+    """ Step data structure """
+    name: Optional[str]
+    how: Optional[str]
+    order: Optional[int]
+
+
+def provides_method(
+        name: str,
+        doc: Optional[str] = None,
+        order: int = DEFAULT_PLUGIN_METHOD_ORDER) -> Any:
+    """
+    A plugin class decorator to register plugin's method with tmt steps.
+
+    In the following example, developer marks ``SomePlugin`` as providing two discover methods,
+    ``foo`` and ``bar``, with ``bar`` being sorted to later position among methods:
+
+    .. code-block:: python
+
+       @tmt.steps.provides_method('foo')
+       @tmt.steps.provides_method('bar', order=80)
+       class SomePlugin(tmt.steps.discover.DicoverPlugin):
+         ...
+
+    :param name: name of the method.
+    :param doc: method documentation. If not specified, docstring of the decorated class is used.
+    :param order: order of the method among other step methods.
+    """
+
+    def _method(cls: Type['BasePlugin']) -> Any:
+        plugin_method = Method(name, class_=cls, doc=doc, order=order)
+
+        cast('BasePlugin', cls.__bases__[0])._supported_methods.append(plugin_method)
+
+        return cls
+
+    return _method
+
+
 class PluginIndex(type):
-    """ Plugin metaclass used to register all available plugins """
+    """
+    Plugin metaclass used to register all available plugins.
+
+    .. deprecated:: 1.17
```

Comment:
Feel free to propose better approach - I wanted to mark the use of metaclass deprecated, and I'm not sure the current deprecation policy of tmt internal APIs is (if any, it seems we're starting to hit "breakage" issues).

Reply:
Good idea. AFAIK we have no deprecation policy per se - API is changed as necessary (as you've seen) so private plugins has to cope with that. Only for customer facing metadata we are very conservative and move ideas breaking things to 2.0.